### PR TITLE
task/TUP-194: Integrate dashboard login with Django user model

### PR DESCRIPTION
**Pull request by jarosenb**
## Overview:
Integrates TUP-UI login with the TUP-CMS user model. The flow works like this: 
1. The user tries to visit a protected route (anything other than `/dashboard/login` for now).
2. They get bounced and redirected to `/dashboard/login` if their `x-tup-token` cookie is missing/invalid.
3. Upon authenticating, they are redirected to the dashboard using `window.location.replace()`, causing the Django authentication background to login their user (or create it if it doesn't exist).

## Related:

- [TUP-194](https://jira.tacc.utexas.edu/browse/TUP-194)

## Changes:
- Refactor `LoginComponent` to use the `window.location` API to redirect instead of the history API. This is necessary for the Django backend to recognize the login event.
- Add a custom authentication backend that checks the user's `x-tup-token` cookie and validates it against the TUP-services API.
- Add a new `RestrictedView` that attempts to authenticate the user and redirects to `/dashboard/login` otherwise. 

## Todo:
- Make redirects work for custom apps other than the dashboard (nothing will break now but it isn't smart about redirecting users to the specific page they were trying to access).

## Testing:

1. Navigate to `localhost:8000/dashboard/` and confirm that you are redirected to a login page.
2. Log in using your TACC credentials and confirm that you are redirected to your profile page.
3. Log out and log back in to check that the auth flow works as expected.

### CODE REVIEW

File: apps/tup-cms/src/apps/dashboard/views.py [hunk 1]
```diff
@@ -1,13 +1,27 @@
...
 from django.http import HttpResponse
 from django.conf import settings
 from django.template import loader
+from django.shortcuts import redirect
 from django.conf import settings
+from django.contrib.auth import authenticate, login
 
 
-def AddedView(request):
+def LoginView(request):
```
Comment (jchuahtacc) [inline apps/tup-cms/src/apps/dashboard/views.py:9]: Is this already covered by Django's built in LoginView?

https://docs.djangoproject.com/en/4.1/topics/auth/default/#django.contrib.auth.views.LoginView

File: apps/tup-cms/src/apps/dashboard/views.py [hunk 2]
```diff
@@ -1,13 +1,27 @@
...
     else:
         template = loader.get_template('dashboard/dashboard.html')
-    return HttpResponse(template.render({}, request))
+    resp = HttpResponse(template.render({'baseUrl': settings.TUP_SERVICES_URL}, request))
+    return resp
+
+def RestrictedView(request):
+    user = authenticate(request)
+    if user is None:
+        return redirect('/dashboard/login')
```
Comment (jchuahtacc) [inline apps/tup-cms/src/apps/dashboard/views.py:20]: Can this be a setting?

File: apps/tup-ui/src/components/auth/LoginComponent/LoginComponent.tsx [hunk 3]
```diff
@@ -58,12 +58,11 @@ const LoginError: React.FC<{ status?: number }> = ({ status }) => {
 };
 
 const LoginComponent: React.FC<LoginProps> = ({ className }) => {
-  const navigate = useNavigate();
   const location = useLocation();
   const from = (location.state as { from?: Location })?.from?.pathname || '/';
   const authCallback = useCallback(() => {
-    navigate(from, { replace: true });
-  }, [from, navigate]);
+    window.location.replace(`/dashboard${from}`);
```
Comment (jchuahtacc) [inline apps/tup-ui/src/components/auth/LoginComponent/LoginComponent.tsx:64]: can this get a query param for `?next_page` which is the Django default via `django.contrib.auth.views.LoginView`? maybe something like:

```
const [ searchParams, setSearchParams ] = useSearchParams();
const next_page? = searchParams.get("next_page");

...
window.location.replace(`${next_page ?? `/dashboard${from}``);
```

File: apps/tup-ui/src/components/auth/LoginComponent/LoginComponent.tsx [hunk 4]
```diff
@@ -59,9 +59,17 @@ const LoginError: React.FC<{ status?: number }> = ({ status }) => {
...
 const LoginComponent: React.FC<LoginProps> = ({ className }) => {
   const location = useLocation();
-  const from = (location.state as { from?: Location })?.from?.pathname || '/';
+  const [searchParams] = useSearchParams();
+
+  // Redirect precedence is 1) last router history location, 2) query string 'from'
+  // parameter, 3) dashboard base route
+  let from = (location.state as { from?: Location })?.from?.pathname ?? '';
+  if (!from) {
+    from = searchParams.get('from') ?? '/dashboard';
```
Comment (jchuahtacc) [inline apps/tup-ui/src/components/auth/LoginComponent/LoginComponent.tsx:68]: The default `django.contrib.auth.REDIRECT_FIELD_NAME` is `next`. Can you test an interaction with the default `@login_required` decorator?

**Review by wesleyboar — CHANGES_REQUESTED** [review]
Please resolve duplicate `tup-cms` resources. I think merging `main` would do that.

**Review by wesleyboar — COMMENTED** [review]
I remove my request for changes.

But, I am unable to test that this works.

<details><summary>What I Tried… <sub>(click to toggle details)</sub></summary>

0. `npm ci`, `nx build core-components`, `nx build core-styles`, `nx serve tup-ui`.
1. Visiting http://localhost:3000/static/ (as `nx serve tup-ui` states) loads white screen and prints to console a warning:
    ```
    <Router basename="/dashboard"> is not able to match the URL "/static/" because it does not start with the basename, so the <Router> won't render anything.
    ```
2. Visiting http://localhost:8000/dashboard/ (as Testing steps state) returns a 404 page with debug info.
3. Visiting http://localhost:3000/dashboard/ (I am grasping for straws) returns a 404 page with:
    > The server is configured with a public base URL of /static/ - did you mean to visit [/static/dashboard/](http://localhost:3000/static/dashboard/) instead?
4. Visiting http://localhost:3000/static/dashboard/ (as 404 page suggests) loads white screen and prints to console a warning:
    ```
    <Router basename="/dashboard"> is not able to match the URL "/static/" because it does not start with the basename, so the <Router> won't render anything.
    ```
    (It looks like step 1.)

</details>

**Review by wesleyboar — COMMENTED** [review]
I am trying to remove my request for changes via this review "Comment".

_**Edit**: This attempt to "un-request" changes did not work 🫤_

**Pull request merged**